rcl_action/src/rcl_action/action_client.c

```diff
@@ -0,0 +1,723 @@
+  impl->action_name = rcutils_strdup(action_name, *((rcutils_allocator_t *)allocator));
+  if (NULL == impl->action_name) {
+    allocator->deallocate(impl, allocator->state);
+    return RCL_RET_BAD_ALLOC;
+  }
+  RCUTILS_LOG_DEBUG_NAMED(
+    ROS_PACKAGE_NAME, "Initializing action goal client");
+  rcl_client_options_t goal_client_options = {
+    .qos = options->goal_service_qos, .allocator = *allocator
+  };
+  rcl_ret_t ret, critical_ret;
+  ret = rcl_action_goal_service_client_init(
+    &impl->goal_client, node, 
+    &type_support->goal_service_type_support,
+    impl->action_name, &goal_client_options);
+  if (RCL_RET_OK == ret) {
+    RCUTILS_LOG_DEBUG_NAMED(
+      ROS_PACKAGE_NAME, "Initializing action cancel client");
+    rcl_client_options_t cancel_client_options = {
+      .qos = options->cancel_service_qos, .allocator = *allocator
+    };
+    ret = rcl_action_cancel_service_client_init(
+      &impl->cancel_client, node,
+      &type_support->cancel_service_type_support,
+      impl->action_name, &cancel_client_options);
+    if (RCL_RET_OK == ret) {
+      RCUTILS_LOG_DEBUG_NAMED(
+        ROS_PACKAGE_NAME, "Action cancel client initialized");
+      rcl_client_options_t result_client_options = {
+        .qos = options->result_service_qos, .allocator = *allocator
+      };
+      RCUTILS_LOG_DEBUG_NAMED(
+        ROS_PACKAGE_NAME, "Initializing action result client");
+      ret = rcl_action_result_client_init(
+        &impl->result_client, node,
+        &type_support->result_service_type_support,
+        impl->action_name, &result_client_options);
+      if (RCL_RET_OK == ret) {
+        RCUTILS_LOG_DEBUG_NAMED(
+          ROS_PACKAGE_NAME, "Action cancel result initialized");
+        rcl_subscription_options_t feedback_subscription_options = {
+          .qos = options->feedback_topic_qos,
+          .ignore_local_publications = false,
+          .allocator = *allocator
+        };
+        RCUTILS_LOG_DEBUG_NAMED(
+          ROS_PACKAGE_NAME, "Initializing action feedback subscription");
+        ret = rcl_action_feedback_subscription_init(
+          &impl->feedback_subscription, node, 
+          &type_support->feedback_topic_type_support,
+          impl->action_name, &feedback_subscription_options);
+        if (RCL_RET_OK == ret) {
+          RCUTILS_LOG_DEBUG_NAMED(
+            ROS_PACKAGE_NAME, "Action feedback subscription initialized");
+          rcl_subscription_options_t status_subscription_options = {
+            .qos = options->status_topic_qos,
+            .ignore_local_publications = false,
+            .allocator = *allocator
+          };
+          RCUTILS_LOG_DEBUG_NAMED(
+            ROS_PACKAGE_NAME, "Initializing action status subscription");
+          ret = rcl_action_status_subscription_init(
+            &impl->status_subscription, node,
+            &type_support->status_topic_type_support,
+            impl->action_name, &status_subscription_options);
+          if (RCL_RET_OK == ret) {
```

Comment:
I think in cases like this `goto error:` is ok to avoid nested `if` statements. No need to change what's here though.

Reply:
I thought about using `goto`, but I went this route to simplify the rollback. Otherwise I would've needed a label for each failure stage (_which things have I initialized so far that need finalization to keep things consistent?_)

Reply:
I don't think a label is needed for each failure stage. It looks like `rcl_action_client_fini()` could be called instead of `fini`-ing everything individually

Reply:
Let's be consistent with rest of the codebase then :). Changed in c5dabd05bc66f9c01431ec5cfe1b86815081c823.